### ISSUE TITLE
Package request: quarto for darwin

### ISSUE THREAD

**Issue by b-rodrigues** [NixOS/nixpkgs#251957]
UPDATE 2023-08-29: The problems are now solved and now quarto builds for x86_64-linux and x86_64-darwin (https://github.com/NixOS/nixpkgs/issues/251957#issuecomment-1696993355)

quarto is currently available for "x86_64-linux" only. I tried to edit the `default.nix` file found [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/quarto/default.nix) to make it work on macOS on my nixpkgs fork. I simply changed the download link to download the macOS binary and the platform on line 74 to "x86_64-darwin" and tried building it but that doesn't seem to work. The building process fails with:

```
error: builder for '/nix/store/qij4mc9nczy04hk8nr99fipcp1wq479d-quarto-1.2.475.drv' failed with exit code 1;
       last 10 log lines:
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeDirectory'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeSignature'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeRequirements-1'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeRequirements'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeDirectory'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeSignature'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeRequirements-1'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeRequirements'
       > tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.cs.CodeDirectory'
       > unpacker produced multiple directories
       For full logs, run 'nix log /nix/store/qij4mc9nczy04hk8nr99fipcp1wq479d-quarto-1.2.475.drv'.
error: 1 dependencies of derivation '/nix/store/inclgcmqnvdhwhpya91nfvp0nr2v1f8b-nix-shell.drv' failed to build
```

Here's my attempt: [link to my fork](https://github.com/b-rodrigues/nixpkgs/blob/bfc4fa2aaf5078c3ae01facaa8c60e38e2a3f257/pkgs/development/libraries/quarto/default.nix). Being a beginner with Nix, I'm asking here for help :smile: But this seems to be related to `tar` that might work differently on macOS than on Linux? When decompressing the archive manually I do see the /bin and /share folders that are then copied during installPhase though.

I'm trying this on a hackintosh VM, as I don't own a Mac, but think it would be good if Quarto would be also available for macOS on nixpkgs.

**Metadata**

* homepage URL: https://quarto.org/
* source URL: https://github.com/quarto-dev/quarto-cli
* license: gpl2+ 
* platforms: linux, darwin

pinging package maintainers

@MrTarantoga @minijackson 

fyi: @jbedo @kupac @philipp-baumann

**Comment by b-rodrigues** [comment]
so `tar tzf` on the archive for darwin returns:

```
tar tzf quarto-1.2.475-macos.tar.gz 
./
./bin/
./share/
and then a long list of files
```

while on linux it returns:

```
quarto-1.2.475/
quarto-1.2.475/bin/
many files and then
quarto-1.2.475/share/

many files
```

so there's a top-level folder for the linux archive, but not for the macOS one, which I guess is the source of the problem.

**Comment by b-rodrigues** [comment]
So following the advice from this old thread: https://releases.nixos.org/nix-dev/2016-June/020541.html

solved the issue, quarto now successfully installs on macOS. Here's the line I added to the `default.nix` file: https://github.com/b-rodrigues/nixpkgs/blob/572a740bfffb56dbea63fda503754107d6c01e6b/pkgs/development/libraries/quarto/default.nix#L24

So now the question becomes: how can I rewrite this expression so that it works of both linux and darwin?

**Comment by b-rodrigues** [comment]
I'm trying to rewrite the expression using if then else statements like this: https://github.com/b-rodrigues/nixpkgs/blob/dca8ea46d50c9e586d4c6ccdced5cd0ea20928dd/pkgs/development/libraries/quarto/default.nix#L27 but this fails with `error syntax error, unexpected ID, expecting ":" or "@".

**Comment by jbedo** [comment]
you can do something like:
```
src = {
    "x86_64-linux" = fetchurl {
       url = "...";
       sha256 = ...;
     };
   
     "x86_64-darwin" = fetchurl {
        ...
      };
    }.${system}
```

**Comment by b-rodrigues** [comment]
thanks, this now works on both linux and macOS (x86_64, I don't have the hardware to test aarch64).

would it be ok to open the PR?

**Comment by b-rodrigues** [comment]
@sagikazarmark I saw that you were also interested in getting quarto for macOS (https://github.com/NixOS/nixpkgs/pull/228984#issuecomment-1536278120), I can only test on a hackintosh VM but if you have access to the hardware and are still interested in getting quarto for macOS, that'd be great!

**Comment by sagikazarmark** [comment]
@b-rodrigues I use quarto on macOS (mostly ARM these days, but I have access to Intel), so yeah, I can test it on macOS.

In fact, I have one that runs on my Mac in my NUR repo: https://github.com/sagikazarmark/nur-packages/blob/master/pkgs/quarto/default.nix

Although it's a somewhat older version. Let me see if I can get #228984 working on darwin.

**Comment by b-rodrigues** [comment]
That's cool, thanks! But did you link to the correct expression? The one you linked to seems to download the linux binary?

**Comment by sagikazarmark** [comment]
It does, but it works all the same.

**Comment by b-rodrigues** [comment]
oh wow I wasn't expecting that! Are you also using some kind of compatibility layer?

**Comment by sagikazarmark** [comment]
Not really. It's just a nodejs app as far as I can tell, so it should run on most platforms just fine. The rest of the dependencies already come from Nix.

**Comment by b-rodrigues** [comment]
Ok so that certainly makes things easier! In that case it would just be a matter of adding the platforms to the expression I guess?

**Comment by b-rodrigues** [comment]
hi @sagikazarmark did you manage to update the package ?

**Comment by sagikazarmark** [comment]
I've updated quarto in my NUR: https://github.com/sagikazarmark/nur-packages/blob/master/pkgs/quarto/default.nix

It works just fine on darwin aarch64

**Comment by b-rodrigues** [comment]
@sagikazarmark that’s great! Since there’s no difference between the packages, could you open the PR to get this latest version of quarto on nixpkgs?

**Comment by minijackson** [comment]
Fixed by #259443
